File: 01_Language/01_Functions/java/src/main/java/com/onevgo/functions/ArrayReverse.java
package com.onevgo.functions;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

public class ArrayReverse {
    public static <E> List<E> arrayReverse(List<E> list) {
        List<E> res = new ArrayList<>(list);
        Collections.reverse(res);
        return res;
    }

    public static void main(String[] args) {
        List<Object> input = Arrays.asList("php", 4.0, Arrays.asList("green", "red"));
        System.out.println(arrayReverse(input));
    }
}

File: 01_Language/01_Functions/js/array_reverse.js
function array_reverse(arr, preserveKeys = false) {
    if (Array.isArray(arr) && !preserveKeys) {
        return arr.slice(0).reverse()
    }
    let res = preserveKeys ? {} : []
    let keys = Object.keys(arr)
    let length = keys.length
    for (let i = length - 1; i >= 0; i--) {
        if (preserveKeys) {
            res[keys[i]] = arr[keys[i]]
        } else {
            res.push(arr[keys[i]])
        }
    }
    return res
}

let input = ["php", 4.0, ["green", "red"]]
console.log(array_reverse(input));
console.log(input)
console.log(array_reverse(input, true));
